mrp_production_replace_component/wizard/wiz_replace_component.py

```diff
@@ -0,0 +1,119 @@
+        domain=[('bom_line_ids', '!=', False)])
+    product_uom_qty = fields.Float(string='Quantity', default=1)
+    product_uom_id = fields.Many2one(
+        comodel_name='uom.uom', string='Unit', readonly=True)
+
+    @api.model
+    def _get_remove_product_ids_domain(self):
+        res = []
+        mo = self._get_manuf_order()
+        if mo:
+            # transient may use falsy value
+            mo.invalidate_cache()
+            res = [('id', 'in', [x.product_id.id for x in mo.move_raw_ids])]
+        return res
+
+    @api.onchange('product_id')
+    def _onchange_product_id(self):
+        self.product_uom_id = self.product_id.uom_id or False
+
+    def apply_replacement(self):
+        self.ensure_one()
+        messages = []
+        if self.remove_product_ids:
+            remove_mess = self._remove_raw_materials()
+            if remove_mess:
+                messages.append(_("Component %s removed" % remove_mess))
+        if self.product_id:
+            self._append_raw_materials()
+            messages.append(_("Component %s added" % self.product_id.name))
+        if messages:
+            mo = self._get_manuf_order()
+            mo.message_post(' - '.join(messages))
+        return {'type': 'ir.actions.act_window_close'}
+
+    def _append_raw_materials(self):
+        self.ensure_one()
+        mo = self._get_manuf_order()
+        if mo:
+            routing = False
+            if mo.routing_id:
+                routing = self.routing_id
+            if routing and routing.location_id:
+                source_location = routing.location_id
+            else:
+                source_location = mo.location_src_id
+            bom_line = self.env['mrp.bom.line'].search(
+                [('product_id', '=', self.product_id.id)], limit=1)
+            if not bom_line:
+                raise UserError(
+                    _('Additional raw material must be defined in a bom'))
+            data = {
+                'sequence': 100,
+                'name': mo.name,
+                'date': mo.date_planned_start,
+                'date_expected': mo.date_planned_start,
+                'picking_type_id': mo.picking_type_id.id,
+                'product_id': self.product_id.id,
+                'product_uom_qty': self.product_uom_qty,
+                'product_uom': self.product_id.uom_id.id,
+                'location_id': source_location.id,
+                'location_dest_id':
+                    self.product_id.property_stock_production.id,
+                'raw_material_production_id': mo.id,
+                'company_id': mo.company_id.id,
+                'price_unit': self.product_id.standard_price,
+                'procure_method': 'make_to_stock',
+                'origin': mo.name,
+                'warehouse_id': source_location.get_warehouse().id,
+                'group_id': mo.procurement_group_id.id,
+                'propagate': mo.propagate,
+                'state': 'confirmed',
```

Comment:
Why not draft and then confirm the move, to follow a more classic process?
I think it would be better. 
- create in draft state
- do mo._adjust_procure_method()
- confirm the created move
So you do the same thing than native mrp odoo